README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
camsRad
=======

[![Travis-CI Build Status](https://travis-ci.org/lukas-rokka/camsRad.svg?branch=master)](https://travis-ci.org/lukas-rokka/camsRad) [![codecov.io](https://codecov.io/github/lukas-rokka/camsRad/coverage.svg?branch=master)](https://codecov.io/github/lukas-rokka/camsRad?branch=master)

`camsRad` is a R client for [CAMS radiation service](http://www.soda-pro.com/web-services/radiation/cams-radiation-service). CAMS radiation service provides time series of global, direct, and diffuse irradiations on horizontal surface, and direct irradiation on normal plane for the actual weather conditions as well as for clear-sky conditions. The geographical coverage is the field-of-view of the Meteosat satellite, roughly speaking Europe, Africa, Atlantic Ocean, Middle East (-66° to 66° in both latitudes and longitudes). The time coverage of data is from 2004-02-01 up to 2 days ago. Data are available with a time step ranging from 15 min to 1 month. Target audience are researchers, developers and consultants in need of high resolution solar radiations time series.

Quick start
-----------

### Install

Dev version from GitHub.

``` r
devtools::install_github("lukas-rokka/camsRad")
```

``` r
library("camsRad")
```

### Authentication

To access the CAMS radiation service you need to register at <http://www.soda-pro.com/web-services/radiation/cams-radiation-service>. The email you use at the registration step will be used for authentication, and need to be set with `cams_set_user()`.

``` r
# Authentication
cams_set_user("your@email.com") # An email registered at soda-pro.com
```

### Example 1

Get hourly CAMS solar data into a R data frame. For the location 60° latitude and 15° longitude, and for period 2016-01-01 to 2016-01-15.

``` r

df <- cams_get_radiation(
  lat=60, lng=15, 
  date_begin="2016-07-01", 
  date_end="2016-07-01")
print(df)
```

### Example 2

Retrieve daily CAMS solar data in netCDF format. You need to have the `ncdf4` package installed.

``` r
library(ncdf4)

filename <- paste0(tempfile(), ".nc")

r <- cams_api(
  60, 15, "2016-06-01", "2016-06-10", 
  format="application/x-netcdf",
  time_step = "P01D",
  filename=filename)

# Access the on disk stored ncdf4 file 
nc <- nc_open(r$response$content)

# list names of available variables
names(nc$var)

# create data.frame with timestamp and global horizontal irradiation and plot it
df <- data.frame(
  timestamp = as.POSIXct(nc$dim$time$vals, "UTC", origin="1970-01-01"),
  GHI = ncvar_get(nc, "GHI"))

plot(df, type="l")

nc_close(nc)
```

Meta
----

-   This package and functions herein are provided as is, without any guarantee.
-   Please note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.
-   Please [report any issues or bugs](https://github.com/lukas-rokka/camsRad/issues).


NEWS.md
# camsRad 0.3.0 (2016-11-07)
=========================
* Compliance with rOpenSci review
### MINOR IMPROVEMENTS
* covr::codecov() badge included
* more testing
* removed dependency to readr package
* return data.frame instead of tibble
* new authentication method
* more examples and improved documentation
* cams_api stops if error in httr calls or errir in returned content

# camsRad 0.2.0 (2016-08-22)
=========================
* Realese for onboarding to rOpenSci
### MINOR IMPROVEMENTS
* Exported functions are documented with examples
* README extended with examples
* vignette added
* Travis-CI added
* Test with testhat added


# camsRad 0.1.0 (2016-08-18)
=========================
* Initial public release


tests/testthat.R
library(testthat)
library(camsRad)

test_check("camsRad")


tests/testthat/test-camsRad.R
context("cams set/get user")

test_that("Setting/getting user authentication", {
  testthat::skip_on_travis() # Conflicts with Travis environment

  username_old=Sys.getenv("CAMS_USERNAME")
  if(username_old=="") {username <- "your@email.com"}

  expect_error(cams_set_user("not a valid email"))
  expect_silent(cams_set_user(username))
  expect_identical(camsRad:::cams_get_user(), username)

  Sys.setenv('CAMS_USERNAME'=username_old)
})


context("cams_api")

test_that("Calling CAMS radiation service API", {
  #skip_on_cran()

  username <- Sys.getenv("CAMS_USERNAME")
  if(username=="") skip("need CAMS_USERNAME to be set")

  r <- cams_api(60, 15, "2016-06-01", "2016-06-10",
                format = "application/csv")

  expect_is(r, "list")
  expect_equal(length(r),2)
  expect_equal(r$ok, TRUE)
  expect_equal(httr::has_content(r$response), TRUE)
  expect_equal(httr::http_type(r$response), "text/csv")
  expect_equal(r$response$status_code, 200)
  expect_gt(length(readLines((rawConnection(r$response$content)))), 250)
})

test_that("Calling cams_get_radiation()", {
  #skip_on_cran()

  username <- Sys.getenv("CAMS_USERNAME")
  if(username=="") skip(" CAMS_USERNAME need to be set")

  df <- cams_get_radiation(60, 15, "2016-06-01", "2016-06-10", time_step="PT01H")

  expect_is(df, "data.frame")
  expect_equal(nrow(df),240)
  expect_equal(ncol(df),11)
  expect_is(df[[1,1]], "POSIXct")
  for(i in 2:11) {expect_is(df[[1,i]], "numeric")}
  expect_true(all(names(df) %in% c("timestamp","TOA","Clear sky GHI",
                                   "Clear sky BHI","Clear sky DHI","Clear sky BNI",
                                   "GHI","BHI","DHI","BNI","Reliability")))
  expect_equal(as.integer(df[[2,1]]) - as.integer(df[[1,1]]), 3600)
})


test_that("Calling cams_get_mcclear()", {
  #skip_on_cran()

  username <- Sys.getenv("CAMS_USERNAME")
  if(username=="") skip("CAMS_USERNAME need to be set")

  df <- cams_get_mcclear(60, 15, "2016-06-01", "2016-06-10", time_step="PT01H")

  expect_is(df, "data.frame")
  expect_equal(nrow(df),240)
  expect_equal(ncol(df),6)
  expect_is(df[[1,1]], "POSIXct")
  for(i in 2:6) {expect_is(df[[1,i]], "numeric")}
  expect_true(all(names(df) %in% c("timestamp","TOA","Clear sky GHI",
                                   "Clear sky BHI","Clear sky DHI","Clear sky BNI")))
  expect_equal(as.integer(df[[2,1]]) - as.integer(df[[1,1]]), 3600)
})


tests/testthat/test-cams_api.R
context("cams_api")

test_that("Calling CAMS radiation service works", {
  #skip_on_cran()

  username=Sys.getenv("CAMS_USERNAME")
  if(username=="") skip("need CAMS_USERNAME to be set")

  r <- cams_api(username, 60, 15, "2016-06-01", "2016-06-10",
                format = "application/csv")

  expect_is(r, "list")
  expect_equal(length(r),2)
  expect_equal(r$ok, TRUE)
  expect_equal(httr::has_content(r$respone), TRUE)
  expect_equal(httr::http_type(r$respone), "text/csv")
  expect_equal(r$respone$status_code, 200)
  expect_gt(length(readr::read_lines((rawToChar(r$respone$content)))), 250)
})


R/cams_api.R
#' API client for
#' \href{http://www.soda-pro.com/web-services/radiation/cams-radiation-service}{CAMS
#' radiation service}
#'
#' @inheritParams cams_get_radiation
#' @param time_ref Time reference:'UT' for universal time, 'TST' for true solar
#'   time. Default 'UT'
#' @param service 'get_mcclear' for CAMS McClear data, 'get_cams_radiation' for
#'   CAMS radiation data. Default 'get_cams_radiation'
#' @param format 'application/csv', 'application/json', 'application/x-netcdf'
#'   or 'text/csv'. Default 'application/csv'
#' @param filename path to file on disk to write to. If empty, data is kept in
#'   memory. Default empty
#'
#' @return list(ok=TRUE/FALSE, response=response). If ok=TRUE, response is the
#'   response from httr::GET. If ok=FALSE, response holds exception text
#'
#' @examples \dontrun{
#' library(ncdf4)
#'
#' filename <- paste0(tempfile(), ".nc")
#'
#' # API call to CAMS
#' r <- cams_api(
#'   60, 15,                       # latitude=60, longitude=15
#'   "2016-06-01", "2016-06-10",   # for 2016-06-01 to 2016-06-10
#'   time_step="PT01H",            # hourly data
#'   service="get_cams_radiation", # CAMS radiation
#'   format="application/x-netcdf",# netCDF format
#'   filename=filename)            # file to save to
#'
#' # Access the on disk stored ncdf4 file
#' nc <- nc_open(r$response$content)
#' # list names of available variables
#' names(nc$var)
#'
#' # create data.frame with timestamp and global horizontal irradiation
#' df <- data.frame(datetime=as.POSIXct(nc$dim$time$vals, "UTC",
#'                                      origin="1970-01-01"),
#'                  GHI = ncvar_get(nc, "GHI"))
#'
#' plot(df, type="l")
#'
#' nc_close(nc)
#' }
#'
#' @export

cams_api <- function(
  lat, lng, date_begin, date_end,
  alt=-999, time_step="PT01H", time_ref="UT", verbose=FALSE,
  service="get_cams_radiation", format='application/csv', filename="") {

  # Stop if username is not provided
  username <- cams_get_user()

  body <- paste0(
    '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>',
    '<wps:Execute service="WPS" version="1.0.0" ',
      'xmlns:ows="http://www.opengis.net/ows/1.1" ',
      'xmlns:xlink="http://www.w3.org/1999/xlink" ',
      'xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" ',
      'xsi:schemaLocation="http://www.opengis.net/wps/1.0.0 ../schemas/wps/1.0.0/wpsExecute_request.xsd" ',
      'xmlns:wps="http://www.opengis.net/wps/1.0.0">',
    '<ows:Identifier>', service, '</ows:Identifier>',
    '<wps:DataInputs>',
    '<wps:Input><ows:Identifier>latitude</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',lat,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>longitude</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',lng,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>altitude</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',alt,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>date_begin</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',date_begin,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>date_end</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',date_end,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>time_ref</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',time_ref,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>summarization</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',time_step,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>verbose</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',"false",'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>username</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',username,'</wps:LiteralData></wps:Data></wps:Input>',
    '</wps:DataInputs>',
    '<wps:ResponseForm>',
    '<wps:ResponseDocument storeExecuteResponse="false">',
    '<wps:Output mimeType="',format,'" asReference="true">',
    '<ows:Identifier>irradiation</ows:Identifier></wps:Output></wps:ResponseDocument>',
    '</wps:ResponseForm>',
    '</wps:Execute>')

  if(verbose) httr::set_config(httr::verbose())

  r <- httr::POST(
    "http://www.soda-is.com/service/wps",
    body=body,
    httr::content_type("text/xml; charset=utf.8"),
    encode= "multipart",
    httr::accept_xml())

  httr::reset_config()

  # stop if request not successful
  httr::stop_for_status(r, "request data from soda-pro.com")

  if (httr::http_type(r) != "application/xml") {
    stop("soda-pro.com did not return xml", call. = FALSE)
  }

  # parse content into xml
  parsed <- xml2::read_xml(
    httr::content(r, as="text", encoding="UTF-8"))

  if(verbose) print(parsed)

  # test if //wps:ProcessSucceeded //wps:Reference exists, else return ExceptionText if exist
  if(length(xml2::xml_find_all(parsed, "//wps:ProcessSucceeded"))==0 |
     length(xml2::xml_find_all(parsed, "//wps:Reference"))==0) {
    if(length(xml2::xml_find_all(parsed, "//ows:ExceptionText"))>0) {
      return(list(
        ok=FALSE,
        response=xml2::xml_text(xml2::xml_find_all(parsed, "//ows:ExceptionText"))))
    } else {
      return(list(ok=FALSE, response=r))
    }
  }

  # get url to the processed file on soda-pro server
  url <- xml2::xml_attr(xml2::xml_find_all(parsed, "//wps:Reference"), "href")

  # if filename=="" the data is saved in memory, oterwise data is writen to disk
  if(filename=="") {
    r <- httr::GET(url)
  } else {
    r <- httr::GET(url, httr::write_disk(filename, overwrite = TRUE))
  }

  return(list(ok=(r$status_code==200), response=r))
}



R/cams_get.R
#' Retrieve CAMS solar radiation data
#'
#' @param lat Latitude, in decimal degrees. Required
#' @param lng Longitude, in decimal degrees. Required
#' @param date_begin Start date as 'yyyy-mm-dd' string. Required
#' @param date_end End date as 'yyyy-mm-dd' string. Required
#' @param alt Altitude in meters, use -999 to let CAMS decide. Default -99
#' @param time_step Aggregation: 'PT01M' for minutes, 'PT15M' for 15 minutes,
#'   'PT01H' for hourly, 'P01D' for daily, 'P01M' for monthly. Deafult 'PT01H'
#' @param verbose TRUE for verbose output. Default "FALSE"
#'
#' @return A data frame with requested solar data
#'
#' @examples \dontrun{
#' # Get hourly solar radiation data
#' df <- cams_get_radiation(
#'   lat=60, lng=15,
#'   date_begin="2016-06-01", date_end="2016-06-15")
#' head(df)
#'
#' # Get daily solar radiation data
#' df <- cams_get_radiation(
#'   lat=60, lng=15,
#'   date_begin="2016-06-01", date_end="2016-06-15",
#'   time_step="P01D")
#' head(df)
#' }
#'
#' @export
#'
cams_get_radiation <- function(
  lat, lng, date_begin, date_end,
  time_step="PT01H", alt=-999, verbose=FALSE) {

  r <- cams_api(lat, lng, date_begin, date_end,
                alt, time_step=time_step,
                verbose=verbose,
                service="get_cams_radiation",
                format="application/csv",
                filename = tempfile())
  if(r$ok==FALSE) stop(r$response, call. = FALSE)
  return(cams_parse(r$response$content))
}

#' Retrieve McClear clear sky solar radiation data
#' @inheritParams cams_get_radiation
#'
#' @return A data frame with requested solar data
#'
#' @examples \dontrun{
#' df <- cams_get_mcclear(
#'   lat=60, lng=15, date_begin="2016-01-01", date_end="2016-01-15")
#' print(head(df))
#' }
#'
#' @export
#'
cams_get_mcclear <- function(
  lat, lng, date_begin, date_end,
  time_step="PT01H", alt=-999, verbose=FALSE) {

  r <- cams_api(lat, lng, date_begin, date_end,
                alt, time_step=time_step,
                verbose=verbose,
                service="get_mcclear",
                format="application/csv",
                filename = tempfile())
  if(r$ok==FALSE) stop(r$response, call. = FALSE)
  return(cams_parse(r$response$content))
}


#' internal parser of csv data.
#' TODO: could break if the csv formating is changed,
#' use json or ncdf instead (more reliable but slower)?
#' @noRd
#'
cams_parse <- function(content) {
  # Last row with '#' holds column names, the data part starts after that
  j <- 0
  for(row_str in readLines(content, n=50)) {
    if(substr(row_str,1,1) != "#") break
    j <- j + 1
  }

  # output info part of the csv file
  message(writeLines(readLines(content, n=j)))

  # get column names, change first column name to 'timestamp'
  col_names <- unlist(strsplit((readLines(content, n=j))[j], ";"))
  col_names <- c("timestamp", col_names[2:(length(col_names))])

  # get data, parse first column as datetime, set column names
  df <- utils::read.delim(content, header=FALSE, sep=";", skip=j)
  df[, 1] <- as.POSIXct(substr(df[, 1], 23, 38), "UTC", "%Y-%m-%dT%H:%M")
  names(df) <- col_names

  return(df)
}


R/cams_auth.R
#' Set username used for authentication by CAMS radiation service
#'
#' @param username Email registered at soda-pro.com. Required
#' @examples \dontrun{
#' # cams_set_user("your@email.com") # An email registered at soda-pro.com
#' }
#'
#' @export
cams_set_user <- function(username) {
  if (!is.character(username) |
      identical(username, "") |
      !grepl("@", username)) {
    stop("Please, provide a valid email registered at CAMS radiation service",
         call. = FALSE)
  }
  Sys.setenv('CAMS_USERNAME'= username)
}

#' internal check if username is given
#'
#' @noRd
cams_get_user <- function() {
  username <- Sys.getenv('CAMS_USERNAME')
  if (identical(username, "")) {
    stop('Provide your email, registered at CAMS radiation service, with cams_set_user("your@email.com")',
         call. = FALSE)
  }
  return(username)
}


R/cams_rad.r
#' API client for
#'   \href{http://www.soda-pro.com/web-services/radiation/cams-radiation-service}{CAMS
#'   radiation service}
#'
#' @param username The email registered at soda-pro.com.
#' @param lat The latitude
#' @param lng The longitude
#' @param date_begin Start date as 'yyy-mm-dd' string
#' @param date_end End date as 'yyy-mm-dd' string
#' @param alt The altitude in meters, use -999 to let CAMS decide
#' @param time_step 'PT01M' for minutes, 'PT15M' for 15 minutes, 'PT01H' for
#'   hourly, 'P01D' for daily, 'P01M' for monthly
#' @param time_ref 'UT' for universal time, 'TST' for true solar time
#' @param verbose TRUE for verbose output
#' @param service 'get_mcclear' for CAMS McClear data, get_cams_radiation for CAMS
#'   radiation data
#' @param format 'application/csv', 'application/json',
#'   'application/x-netcdf' or 'text/csv'
#' @param filename path to file on disk to write to. If empty, data is kept in memory.
#'
#' @return list(ok=TRUE/FALSE, response=response). If ok=TRUE, response is the
#'   response from httr::GET. If ok=FALSE, response holds exception text
#'
#' @examples \dontrun{
#' library(ncdf4)
#'
#' filename <- paste0(tempfile(), ".nc")
#'
#' r <- cams_api(username, 60, 15, "2016-06-01", "2016-06-10",
#'               format = "application/x-netcdf", filename=filename)
#'
#' # Access the on disk stored ncdf4 file
#' nc <- nc_open(r$respone$content)
#' # list names of available variables
#' names(nc$var)
#'
#' # create data.frame with timestamp and global horizontal irradiation
#' df <- data.frame(datetime=as.POSIXct(nc$dim$time$vals, "UTC",
#'                                      origin="1970-01-01"),
#'                  GHI = ncvar_get(nc, "GHI"))
#'
#' plot(df, type="l")
#'
#' nc_close(nc)
#' }
#'
#' @import dplyr
#' @export

cams_api <- function(username, lat, lng, date_begin, date_end,
                     alt=-999, time_step="PT01H", time_ref="UT", verbose=FALSE,
                     service="get_cams_radiation", format='application/csv', filename="") {
  body <- paste0(
    '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>',
    '<wps:Execute service="WPS" version="1.0.0" ',
      'xmlns:ows="http://www.opengis.net/ows/1.1" ',
      'xmlns:xlink="http://www.w3.org/1999/xlink" ',
      'xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" ',
      'xsi:schemaLocation="http://www.opengis.net/wps/1.0.0 ../schemas/wps/1.0.0/wpsExecute_request.xsd" ',
      'xmlns:wps="http://www.opengis.net/wps/1.0.0">',
    '<ows:Identifier>', service, '</ows:Identifier>',
    '<wps:DataInputs>',
    '<wps:Input><ows:Identifier>latitude</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',lat,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>longitude</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',lng,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>altitude</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',alt,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>date_begin</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',date_begin,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>date_end</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',date_end,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>time_ref</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',time_ref,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>summarization</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',time_step,'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>verbose</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',"false",'</wps:LiteralData></wps:Data></wps:Input>',
    '<wps:Input><ows:Identifier>username</ows:Identifier>',
    '<wps:Data><wps:LiteralData>',username,'</wps:LiteralData></wps:Data></wps:Input>',
    '</wps:DataInputs>',
    '<wps:ResponseForm>',
    '<wps:ResponseDocument storeExecuteResponse="false">',
    '<wps:Output mimeType="',format,'" asReference="true">',
    '<ows:Identifier>irradiation</ows:Identifier></wps:Output></wps:ResponseDocument>',
    '</wps:ResponseForm>',
    '</wps:Execute>')

  if(verbose) httr::set_config(httr::verbose())

  r <- tryCatch(
    httr::POST("http://www.soda-is.com/service/wps",
               body=body,
               httr::content_type("text/xml; charset=utf.8"),
               encode= "multipart",
               httr::accept_xml())
    , error = function(e) {e})

  if(is.null(r$status_code)) {
    # should not happen
    stop(r)
  } else if(r$status_code !=200) {
    stop(httr::content(r))
  }

  parsed <- httr::content(r)

  if(verbose) print(parsed)

  # test if //wps:ProcessSucceeded //wps:Reference exists, else return ExceptionText if exist
  if(length(xml2::xml_find_all(parsed, "//wps:ProcessSucceeded"))==0 |
     length(xml2::xml_find_all(parsed, "//wps:Reference"))==0) {
    httr::reset_config()
    if(length(xml2::xml_find_all(parsed, "//ows:ExceptionText"))>0) {
      return(list(ok=FALSE,
                  response=xml2::xml_text(xml2::xml_find_all(parsed, "//ows:ExceptionText"))))
    } else {
      return(list(ok=FALSE, response=r))
    }
  }

  # get url to the processed file on soda-pro server
  url <- xml2::xml_attr(xml2::xml_find_all(parsed, "//wps:Reference"), "href")

  # if filename=="" the data is saved in memory, oterwise data is writen to disk
  if(filename=="") {
    r <- httr::GET(url)
  } else {
    r <- httr::GET(url, httr::write_disk(filename, overwrite = TRUE))
  }
  httr::reset_config()
  return(list(ok=(r$status_code==200), respone=r))
}

#' Retrieve CAMS solar radiation data
#' @inheritParams cams_api
#'
#' @return A tibble (data frame) with requested solar data
#'
#' @examples \dontrun{
#' username <- "your@email.com" # An email registered at soda-pro.com
#' df <- cams_get_radition(username, lat=60, lng=15,
#'                         date_begin="2016-01-01", date_end="2016-01-15")
#' print(df)
#' }
#'
#' @export

cams_get_radition <- function(username, lat, lng, date_begin, date_end,
                              time_step="PT01H", alt=-999, verbose=FALSE) {
  r <- cams_api(username, lat, lng, date_begin, date_end, alt, time_step=time_step,
                verbose=verbose, service="get_cams_radiation", format="application/csv")
  if(r$ok==FALSE) {
    stop(r$response, call. = FALSE)
  }
  df <- cams_parse(readr::read_file(r$respone$content))
  return(df)
}

#' Retrieve McClear clear sky solar radiation data
#' @inheritParams cams_api
#'
#' @return A tibble (data.frame) with requested solar data
#'
#' @examples \dontrun{
#' username <- "your@email.com" # An email registered at soda-pro.com
#' df <- cams_get_mcclear(username, lat=60, lng=15,
#'                        date_begin="2016-01-01", date_end="2016-01-15")
#' print(df)
#' }
#'
#' @export
#'
cams_get_mcclear <- function(username, lat, lng, date_begin, date_end,
                             time_step="PT01H", alt=-999, verbose=FALSE) {
  r <- cams_api(username, lat, lng, date_begin, date_end, alt, time_step=time_step,
                verbose=verbose, service="get_mcclear", format="application/csv")
  if(r$ok==FALSE) {
    stop(r$response, call. = FALSE)
  }
  df <- cams_parse(readr::read_file(r$respone$content))
  return(df)
}

#' internal parser of csv data.
#' TODO: could break if the csv formating is changed,
#' use json or ncdf instead (more reliable but slower)?
#' @noRd
cams_parse <- function(r_str) {
  # Last row with '#' holds column names, the data part starts after that
  j <- 0
  for(row_str in readr::read_lines(r_str, n_max=50)) {
    if(substr(row_str,1,1) != "#") break
    j <- j + 1
  }

  # output info part of the csv file
  message(writeLines(readr::read_lines(r_str, n_max=j)))

  # get column names, change first column name to 'timestamp'
  col_names <- unlist(readr::read_lines(r_str, skip=j-1, n_max=1) %>% strsplit(";"))
  col_names <- c("timestamp", col_names[2:(length(col_names))])

  # get data, parse first column as datetime
  df <- readr::read_delim(r_str, ";", col_names = FALSE, skip=j) %>%
    mutate(X1=readr::parse_datetime(paste(substr(X1,23,32), substr(X1,34,38)))) %>%
    stats::setNames(col_names)

  return(df)
}
